### ISSUE TITLE
Problems with the task cancellation for requests

### ISSUE THREAD

**Issue by MrKang** [Nomi-ngj/NetworkMoyaAsync#1]
Hi, I read your code but I didn't find any about moya `cancellable.cancel()` .The swift task cancellation won't automatically cancel your moya request, you need to control it manually. I have tried to do this, but it seems I can not using continutation to implement 
truly async moya💔

**Comment by Nomi-ngj** [comment]
```
private var fetchUsersTask: Task<Void, Never>? = nil // Task to manage ongoing fetch user requests
fetchUsersTask?.cancel()

// Create a new task for fetching users
fetchUsersTask = Task {
    // Fetch users logic here
}
```

Does this approach not work for you?
No worries—I’ll make some adjustments to help streamline it.

**Comment by Nomi-ngj** [comment]
Hi MrKang,

Thank you for your feedback! I've resolved the cancellation issue in the code. Now, you can cancel the ongoing Moya request when the user initiates it.

### How to Use Cancellable
Here’s a brief overview of how the Cancellable is implemented and how you can use it:

```
let concurrency = MoyaConcurrency(provider: provider)

let task = Task {
    do {
        let users: [User] = try await concurrency.request(.user(request: userRequest))
        // Handle success
    } catch {
        // Handle cancellation or other errors
        if error is NetworkError && (error as! NetworkError == .requestCancelled) {
            print("Request was cancelled by the user.")
        } else {
            print("An error occurred: \(error.localizedDescription)")
        }
    }
}

// When the user wants to cancel the request:
concurrency.cancelCurrentRequest() // This will cancel the ongoing request.
```


**Creating a Request:** You can initiate a request using the request method, which handles the network call asynchronously. For example:

`let users: [User] = try await provider.async.request(.user(request: userRequest))
`

**Cancelling the Request:** If the user decides to cancel the ongoing request, you can call the cancelCurrentRequest() method from the MoyaConcurrency instance. This will cancel the current Moya request and handle it gracefully:

`concurrency.cancelCurrentRequest()`

**Handling Cancellation:** Inside the request closure, you can check if the request has been cancelled. If it has, you can manage the cancellation as needed:

```
if self.cancellableRequest?.isCancelled ?? false {
    continuation.resume(throwing: NetworkError.requestCancelled)
    return
}
```

Summary
With this approach, you can easily manage ongoing requests and provide users with the ability to cancel them when needed. Feel free to reach out if you have any further questions or need more clarification!

Best,
Nouman Gul

**Comment by MrKang** [comment]
Hi @Nomi-ngj 

I'm sorry, I haven't checked my email list of unread messages lately. Your MoyaConcurrency wrapper is interesting, using moya's cancelable to cancel the request directly thus bypassing the task's cancel(). I still haven't given up trying lately, and I have found a way to cancel the request via task's cancel(), but I haven't done any testing yet, so I can show you my code, which may help you.

## wrapping Cancellable

```swift
final class CancellableWrapper: @unchecked Sendable {
    var cancellable: Cancellable?
}
```

## using in continuation closure

```swift
func request<T: Decodable>(_ target: Target) async throws -> Task<T, Error> {
    return Task { in
           let cancelWrapper = CancellableWrapper()
           return try await withTaskCancellationHandler {
            return try await withCheckedThrowingContinuation { continuation in
                // Cancel any existing request before starting a new one
                
                
                // Start the Moya request
                cancelWrapper.cancellable =  provider.request(target) { [weak self] result in
                    
                    guard let self else { return }
                    
                    // Check if the request was cancelled
                    if self.cancellableRequest?.isCancelled ?? false {
                        continuation.resume(throwing: NetworkError.requestCancelled)
                        return
                    }
                    
                    switch result {
                    case .success(let response):
                        // Attempt to decode the response into the specified type T
                        guard let res = try? JSONDecoder.default.decode(T.self, from: response.data) else {
                            continuation.resume(throwing: NetworkError.parsingError)
                            return
                        }
                        // Resume with the decoded result
                        continuation.resume(returning: res)
                        
                    case .failure(let error):
                        // Resume with the encountered error
                        continuation.resume(throwing: error)
                    }
                }
            }
        } onCancel: { cancelWrapper.cancellable.cancel() }
    }
}
```

The request func is base on your code of `MoyaConcurrency`. It's just pseudo-code and maybe a little bit of a problem, but the idea seems to be the right one, I've tried it and it does cancel the request via task.cancel
